README.md
# lcs-image-diff
Image diff library and tool with LCS algorithm. rust port of [murooka/go-diff-image](https://github.com/murooka/go-diff-image)

[![](http://meritbadge.herokuapp.com/lcs-image-diff)](https://crates.io/crates/lcs-image-diff)
[![CircleCI](https://circleci.com/gh/bokuweb/lcs-image-diff-rs/tree/master.svg?style=svg)](https://circleci.com/gh/bokuweb/lcs-image-diff-rs/tree/master)

## Requirements
- latest Rust (recommend [rustup](https://www.rustup.rs/))

## Library

### Usage

```toml
# Cargo.toml
[dependencies]
image = "0.20"
lcs-image-diff = { version = "0.1", default-features = false }
```

```rust
use lcs_image_diff::compare;

let mut before = image::open("before.png")?;
let mut after = image::open("after.png")?;

let diff = compare(&mut before, &mut after, 100.0 / 256.0)?;

before.save("marked_before.png")?;
after.save("marked_after.png")?;
diff.save("diff.png")?;
```

## Binary

### Installation

``` bash
cargo install lcs-image-diff
```

### Usage

```
lcs-image-diff path/to/before.png path/to/after.png path/to/diff.png
```

## Example

| before.png        | after.png          | diff.png                 |
| --------------- |---------------| -------------------- |
| ![](https://github.com/bokuweb/lcs-image-diff-rs/blob/master/test/images/before.png?raw=true) | ![](https://github.com/bokuweb/lcs-image-diff-rs/blob/master/test/images/after.png?raw=true) |![](https://github.com/bokuweb/lcs-image-diff-rs/blob/master/test/images/diff.png?raw=true)|

`lcs-image-diff` outputs marked before and after images too.

| marked_before.png        | marked_after.png          |
| --------------- |---------------|
| ![](https://github.com/bokuweb/lcs-image-diff-rs/blob/master/test/images/marked_before.png?raw=true) | ![](https://github.com/bokuweb/lcs-image-diff-rs/blob/master/test/images/marked_after.png?raw=true) |

## LICENSE

The MIT License (MIT)

Copyright (c) 2018 @bokuweb

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# CHANGELOG


<a name="0.1.7"></a>
## [0.1.7](https://github.com/bokuweb/lcs-image-diff-rs/compare/0.1.6...0.1.7) (2019-01-06)

### Refactor

Split crate to library and binary #4

<a name="0.1.6"></a>
## [0.1.6](https://github.com/bokuweb/lcs-image-diff-rs/compare/0.1.5...0.1.6) (2018-03-15)

### Bug Fixes

* Fixed a bug, panic occurred when difference width image input

<a name="0.1.5"></a>
## [0.1.5](https://github.com/bokuweb/lcs-image-diff-rs/compare/0.1.4...0.1.5) (2018-02-25)

### Perf

* Improve performance

<a name="0.1.4"></a>
## [0.1.4](https://github.com/bokuweb/lcs-image-diff-rs/compare/0.1.3...0.1.4) (2018-02-23)

### Bug Fixes

* Fix bug, when specify only filename permission required


<a name="0.1.3"></a>
## [0.1.3](https://github.com/bokuweb/lcs-image-diff-rs/compare/0.1.2...0.1.3) (2018-02-23)

### Bug Fixes

* Fix bug, panic occurred when dir is already exists

<a name="0.1.1"></a>
## [0.1.1](https://github.com/bokuweb/lcs-image-diff-rs/compare/0.1.0...0.1.1) (2018-02-23)

### Bug Fixes

* Fix marked range size was wrong.

<a name="0.1.2"></a>
## [0.1.2](https://github.com/bokuweb/lcs-image-diff-rs/compare/0.1.1...0.1.2) (2018-02-23)

### Features

* Modify default marked color

<a name="0.1.1"></a>
## [0.1.1](https://github.com/bokuweb/lcs-image-diff-rs/compare/0.1.0...0.1.1) (2018-02-23)

### Bug Fixes

* Fix marked range size was wrong.

### Features

* Modify default marked color


<a name="0.1.0"></a>
## 0.1.0 (2018-02-13)

First release :tada:


src/main.rs
extern crate futures;
extern crate futures_cpupool;
extern crate image;
extern crate lcs_image_diff;

#[macro_use]
extern crate clap;

mod mkdir;
mod rename;

use std::path::Path;
use image::*;
use rename::*;
use mkdir::*;
use clap::{App, Arg};
use futures_cpupool::CpuPool;
use futures::{future, Future};
use lcs_image_diff::compare;

static RATE: f32 = 100.0 / 256.0;

pub fn save_image(image: &DynamicImage, filename: &str) {
    let path = Path::new(filename).parent().unwrap();
    let _result = mkdirp(path);
    image.save(filename).unwrap();
}

fn main() {
    let app = App::new(crate_name!())
        .version(crate_version!())
        .author(crate_authors!())
        .about(crate_description!())
        .arg(
            Arg::with_name("before_image")
                .help("path to before image")
                .required(true),
        )
        .arg(
            Arg::with_name("after_image")
                .help("path to after image")
                .required(true),
        )
        .arg(
            Arg::with_name("diff_image")
                .help("path to diff image")
                .required(true),
        );
    let matches = app.get_matches();
    let before_image = matches.value_of("before_image").unwrap();
    let after_image = matches.value_of("after_image").unwrap();
    let diff_image = matches.value_of("diff_image").unwrap().to_owned();
    let mut before = image::open(before_image).unwrap();
    let mut after = image::open(after_image).unwrap();

    let marked_before = add_prefix_to_file_name(&before_image, &"marked_");
    let marked_after = add_prefix_to_file_name(&after_image, &"marked_");

    let result = compare(&mut before, &mut after, RATE).unwrap();

    {
        let thread_pool = CpuPool::new_num_cpus();
        let before_thread = thread_pool.spawn_fn(move || -> Result<(), ()> {
            save_image(&before, &marked_before);
            Ok(())
        });
        let after_thread = thread_pool.spawn_fn(move || -> Result<(), ()> {
            save_image(&after, &marked_after);
            Ok(())
        });
        let result_thread = thread_pool.spawn_fn(move || -> Result<(), ()> {
            save_image(&result, &diff_image);
            Ok(())
        });
        future::join_all(vec![before_thread, after_thread, result_thread])
            .wait()
            .unwrap();
    }
}


src/rename.rs
use std::path::Path;

pub fn add_prefix_to_file_name(file_name: &str, prefix: &str) -> String {
    let path = Path::new(file_name);
    let file_name = path.file_name().unwrap();
    let dir = path.parent().unwrap();
    if dir.to_str().unwrap() == "" {
        return format!("{}{}", prefix, file_name.to_str().unwrap());
    }
    format!("{}/{}{}",
            dir.to_str().unwrap(),
            prefix,
            file_name.to_str().unwrap())
}

#[test]
fn shoud_add_prefix_with_dir() {
    let filename = add_prefix_to_file_name(&"foo/bar/filename.png", &"sample_");
    assert_eq!(filename, "foo/bar/sample_filename.png");
}

#[test]
fn shoud_add_prefix_without_dir() {
    let filename = add_prefix_to_file_name(&"filename.png", &"sample_");
    assert_eq!(filename, "sample_filename.png");
}

src/image_creator.rs
use std::cmp;
use super::image::*;
use super::lcs_diff::*;
use super::base64::decode;

pub static BLACK: (u8, u8, u8) = (0, 0, 0);
pub static RED: (u8, u8, u8) = (255, 119, 119);
pub static GREEN: (u8, u8, u8) = (99, 195, 99);

fn compute_range(r: &Vec<usize>) -> Vec<(usize, usize)> {
    let mut i = 0;
    let mut j = 0;
    let mut acc: usize;
    let mut y1: usize;
    let mut ranges: Vec<(usize, usize)> = Vec::new();
    while i < r.len() {
        y1 = r[i];
        acc = y1;
        i += 1;
        loop {
            if i >= r.len() {
                break;
            }
            let index = r[i];
            if acc + 1 != index {
                break;
            }
            acc = index;
            i += 1;
            j += 1;
        }
        let y2 = y1 + j;
        j = 0;
        ranges.push((y1, y2));
    }
    ranges
}

fn blend_diff_area<G>(img: &mut G, ranges: Vec<(usize, usize)>, rgb: (u8, u8, u8), rate: f32)
where
    G: GenericImage<Pixel = Rgba<u8>>,
{
    for (y1, y2) in ranges {
        for y in y1..(y2 + 1) {
            for x in 0..img.dimensions().0 {
                let p = img.get_pixel(x, y as u32);
                let blended = blend(p, rgb, rate);
                img.put_pixel(x, y as u32, blended);
            }
        }
    }
}

fn blend(base: Rgba<u8>, rgb: (u8, u8, u8), rate: f32) -> Rgba<u8> {
    return Rgba {
        data: [
            (base.data[0] as f32 * (1.0 - rate) + rgb.0 as f32 * (rate)) as u8,
            (base.data[1] as f32 * (1.0 - rate) + rgb.1 as f32 * (rate)) as u8,
            (base.data[2] as f32 * (1.0 - rate) + rgb.2 as f32 * (rate)) as u8,
            base.data[3],
        ],
    };
}

fn put_diff_pixels(
    y: usize,
    img: &mut ImageBuffer<Rgba<u8>, Vec<u8>>,
    row_width: u32,
    data: &String,
    rgb: (u8, u8, u8),
    rate: f32,
) -> Result<(), base64::DecodeError> {
    let row = decode(data)?;
    for x in 0..img.dimensions().0 {
        let index = x as usize * 4;
        let pixel: Rgba<u8> = if row_width > x {
            Rgba {
                data: [row[index], row[index + 1], row[index + 2], row[index + 3]],
            }
        } else {
            Rgba { data: [0, 0, 0, 0] }
        };
        img.put_pixel(x as u32, y as u32, blend(pixel, rgb, rate));
    }
    Ok(())
}

pub fn mark_org_image(
    base: &mut DynamicImage,
    color: (u8, u8, u8),
    rate: f32,
    indexes: &Vec<usize>,
) {
    let range = compute_range(indexes);
    blend_diff_area(base, range, color, rate);
}

pub fn get_diff_image(
    before_width: u32,
    after_width: u32,
    result: &Vec<DiffResult<String>>,
    rate: f32,
) -> Result<DynamicImage, base64::DecodeError> {
    let height = result.len() as u32;
    let width = cmp::max(before_width, after_width);
    let mut img: ImageBuffer<Rgba<u8>, Vec<u8>> = ImageBuffer::new(width, height);
    for (y, d) in result.iter().enumerate() {
        match d {
            &DiffResult::Added(ref a) => {
                put_diff_pixels(y, &mut img, after_width, &a.data, GREEN, rate)?
            }
            &DiffResult::Removed(ref r) => {
                put_diff_pixels(y, &mut img, before_width, &r.data, RED, rate)?
            }
            &DiffResult::Common(ref c) => {
                put_diff_pixels(y, &mut img, width, &c.data, BLACK, 0.0)?
            }
        }
    }
    Ok(ImageRgba8(img))
}


src/lib.rs
extern crate base64;
extern crate image;
extern crate lcs_diff;

mod diff;
mod image_creator;

use diff::*;
use image_creator::*;
use image::*;
pub use base64::DecodeError;

/// Accepts two mutable references to `image::DynamicImage` and rate.
/// Returns diff `image::DynamicImage` and marks removed and added
/// parts on input images.
///
/// # Examples
///
/// ```no_run
/// extern crate image;
/// # use std::error::Error;
/// # fn main() -> Result<(), Box<Error>> {
/// use lcs_image_diff::compare;
///
/// let mut before = image::open("before.png")?;
/// let mut after = image::open("after.png")?;
///
/// let diff = compare(&mut before, &mut after, 100.0 / 256.0)?;
///
/// before.save("marked_before.png")?;
/// after.save("marked_after.png")?;
/// diff.save("diff.png")?;
/// # Ok(())
/// # }
/// ```
pub fn compare(
    before: &mut DynamicImage,
    after: &mut DynamicImage,
    rate: f32
) -> Result<DynamicImage, DecodeError> {
    let compare_before = CompareImage::new(before.dimensions(), before.raw_pixels());
    let compare_after = CompareImage::new(after.dimensions(), after.raw_pixels());
    let result = diff(compare_before, compare_after);

    let mut added: Vec<usize> = Vec::new();
    let mut removed: Vec<usize> = Vec::new();
    for d in result.iter() {
        match d {
            &lcs_diff::DiffResult::Added(ref a) => added.push(a.new_index.unwrap()),
            &lcs_diff::DiffResult::Removed(ref r) => removed.push(r.old_index.unwrap()),
            _ => (),
        }
    }

    mark_org_image(before, RED, rate, &removed);
    mark_org_image(after, GREEN, rate, &added);

    get_diff_image(before.dimensions().0, after.dimensions().0, &result, rate)
}
